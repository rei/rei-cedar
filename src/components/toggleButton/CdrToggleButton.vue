<script>
import {
  useCssModule, computed, defineComponent, inject,
} from 'vue';

export default defineComponent({
  name: 'CdrToggleButton',
  props: {
    toggleValue: {
      type: String,
      required: true,
    },
  },
  setup(props) {
    const selectedToggle = inject('selectedToggleValue');
    const isActive = computed(() => props.toggleValue === selectedToggle.value);
    return {
      style: useCssModule(),
      isActive,
    };
  },
});
</script>

<template>
  <li
    role="presentation"
    :class="style['cdr-toggle-button__container']"
  >
    <button
      role="radio"
      :class="style['cdr-toggle-button__item']"
      :value="toggleValue"
      :aria-label="toggleValue"
      :aria-checked="isActive ? 'true' : 'false'"
      :tabindex="isActive ? 0 : -1"
    >
      <slot>{{ toggleValue }}</slot>

    </button>
  </li>
</template>

<style lang="scss" module src="./styles/CdrToggleButton.module.scss">
</style>
