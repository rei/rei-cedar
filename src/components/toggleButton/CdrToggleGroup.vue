<script>
import {
  ref, watch, onMounted, provide, computed, defineComponent, useCssModule,
} from 'vue';
import mapClasses from '../../utils/mapClasses';
import propValidator from '../../utils/propValidator';

export default defineComponent({
  name: 'CdrToggleGroup',
  props: {
    /** @ignore */
    modelValue: {
      type: [String, Number, Boolean, Object, Array],
      required: true,
    },
    /**
     * Sets toggle button size
     * @demoSelectMultiple false
     * @values medium, large
    */
    size: {
      type: String,
      default: 'medium',
      validator: (value) => propValidator(
        value,
        ['medium', 'large'],
      ),
    },
  },
  emits: {
    /**
     * Event emitted by v-model
     * @param modelValue
     */
    'update:modelValue': null,
  },
  setup(props, ctx) {
    const toggleGroup = ref(null);

    const selectedToggleValue = ref(props.modelValue);
    provide('selectedToggleValue', selectedToggleValue);

    const sizeClass = computed(() => (props.size
      ? `cdr-toggle-group--${props.size}`
      : 'cdr-toggle-group--medium'));

    let toggleButtonElements;

    onMounted(() => {
      toggleButtonElements = Array.from(toggleGroup.value.querySelectorAll('button'));
    });

    watch(() => props.modelValue, (value) => {
      selectedToggleValue.value = value;
    });

    const selectToggleButton = (e) => {
      if (!e.target.closest('button')) {
        return;
      }
      const button = e.target.closest('button');
      selectedToggleValue.value = button.value;
      ctx.emit('update:modelValue', button.value);
      button.focus();
    };

    const focusNext = (e) => {
      const currentButton = e.target;
      const currentButtonIndex = toggleButtonElements.indexOf(currentButton);
      const isLastButton = (currentButtonIndex === toggleButtonElements.length - 1);

      if (isLastButton) {
        return;
      }

      const nextButton = toggleButtonElements[currentButtonIndex + 1];
      nextButton.focus();
    };

    const focusPrev = (e) => {
      const currentButton = e.target;
      const currentButtonIndex = toggleButtonElements.indexOf(currentButton);
      const isFirstButton = (currentButtonIndex === 0);

      if (isFirstButton) {
        return;
      }

      const nextButton = toggleButtonElements[currentButtonIndex - 1];
      nextButton.focus();
    };

    const baseClass = 'cdr-toggle-group';
    const style = useCssModule();

    return {
      toggleGroup,
      selectedToggleValue,
      toggleButtonElements,
      selectToggleButton,
      focusNext,
      focusPrev,
      sizeClass,
      baseClass,
      mapClasses,
      style,
    };
  },
});
</script>

<template>
  <ul
    ref="toggleGroup"
    role="radiogroup"
    v-bind="$attrs"
    :class="mapClasses(
      style,
      baseClass,
      sizeClass,
    )"
    @click.prevent="selectToggleButton"
    @keyup.right.prevent="focusNext"
    @keyup.left.prevent="focusPrev"
  >
    <!-- @slot CdrToggleGroup content (CdrToggleButton components) -->
    <slot />
  </ul>
</template>

<style lang="scss" module src="./styles/CdrToggleGroup.module.scss">
</style>
