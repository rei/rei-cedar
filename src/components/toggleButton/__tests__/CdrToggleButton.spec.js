import { afterEach, beforeEach, describe, expect, it } from 'vitest';
import { shallowMount, mount } from '../../../../test/vue-jest-style-workaround.js';
import CdrToggleGroup from '../CdrToggleGroup.vue';
import { h } from 'vue';
import CdrToggleButton from '../CdrToggleButton.vue';

describe('CdrToggleButton', () => {
  let wrapper;
  let firstButton;
  let secondButton;
  let thirdButton;
  describe('component snapshot', () => {
    beforeEach(() => {
      wrapper = mount(CdrToggleGroup, {
        props: {
          modelValue: 'thing-1'
        },
        slots: {
          default: [
            h(CdrToggleButton, { 'toggle-value': 'thing-1' }),
            h(CdrToggleButton, { 'toggle-value': 'thing-2' }),
            h(CdrToggleButton, { 'toggle-value': 'thing-3' }),
          ],
        },
        attachTo: document.body
      });
      firstButton = wrapper.findAll('button')[0];
      secondButton = wrapper.findAll('button')[1];
      thirdButton = wrapper.findAll('button')[2];
      firstButton.element.focus();
    })

    it('matches the snapshot', () => {
      expect(wrapper.element).toMatchSnapshot();
    });

    it('the first button is focused', () => {
      expect(document.activeElement).toBe(firstButton.wrapperElement);
    });

    it('the tabindex of the first button is 0', () => {
      expect(firstButton.attributes('tabindex')).toBe('0');
    })

    it('the tabindex of the second button is -1', () => {
      expect(secondButton.attributes('tabindex')).toBe('-1');
    });

    it('the aria-checked value of the first button is true', () => {
      expect(firstButton.attributes('aria-checked')).toBe('true');
    })

    it('the aria-checked value of the second button is false', () => {
      expect(secondButton.attributes('aria-checked')).toBe('false');
    });

    describe('adding the large prop', ()=>{
      beforeEach(()=>{
        wrapper.setProps({ size: 'large'})
      })

      it('matches the snapshot', () => {
        expect(wrapper.element).toMatchSnapshot();
      });
    })

    describe('changing the model value', () => {
      beforeEach(async()=>{
        wrapper.setProps({ modelValue: 'thing-3'})
      })

      it('should select the expected button', ()=>{
        expect(thirdButton.attributes('aria-checked')).toBe('true');
      })
    });

    describe('when hitting the right arrow key', () => {
      beforeEach(() => {
        firstButton.trigger('keyup.right');
      });
      afterEach(() => {
        firstButton.element.focus();
      });

      it('the second button is focused', () => {
        expect(document.activeElement).toBe(secondButton.wrapperElement);
      })
    })

    describe('when hitting the right arrow key', () => {
      beforeEach(async () => {
        thirdButton.element.focus();
        await wrapper.vm.$nextTick();
        thirdButton.trigger('keyup.left');
      });

      afterEach(() => {
        thirdButton.element.focus();
      });

      it('the second button is focused', () => {
        expect(document.activeElement).toBe(secondButton.wrapperElement);
      })
    });

    //Test for when the third button is selected and you're hitting right

    describe('when clicking on an unchecked button', () => {
      beforeEach(async () => {
        secondButton.trigger('click');
        await wrapper.vm.$nextTick();
      });

      it('matches snapshot', () => {
        expect(wrapper.element).toMatchSnapshot();
      });

      it('turns the aria-checked value to true', () => {
        expect(secondButton.attributes('aria-checked')).toBe('true');
      });

      it('turns the aria-checked value of the previously checked button to false', () => {
        expect(firstButton.attributes('aria-checked')).toBe('false');
      });

      it('the tabindex of the first button is -1', () => {
        expect(firstButton.attributes('tabindex')).toBe('-1');
      });

      it('the tabindex of the second button is 0', () => {
        expect(secondButton.attributes('tabindex')).toBe('0');
      });

      it('the second button is focused', () => {
        expect(document.activeElement).toBe(secondButton.wrapperElement);
      });

      it('updates the modelValue with the expected value', () => {
        expect(wrapper.emitted()['update:modelValue'][0][0]).toBe('thing-2');
      })
    });
  })
});
