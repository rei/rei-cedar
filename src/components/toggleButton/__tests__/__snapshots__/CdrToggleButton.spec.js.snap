// Vitest Snapshot v1

exports[`CdrToggleButton > component snapshot > adding the large prop > matches the snapshot 1`] = `
<ul
  class="cdr-toggle-group cdr-toggle-group--large"
  role="radiogroup"
>
  
  <li
    class="cdr-toggle-button__container"
    role="presentation"
  >
    <button
      aria-checked="true"
      aria-label="thing-1"
      class="cdr-toggle-button__item"
      role="radio"
      tabindex="0"
      value="thing-1"
    >
      
      thing-1
      
    </button>
  </li>
  <li
    class="cdr-toggle-button__container"
    role="presentation"
  >
    <button
      aria-checked="false"
      aria-label="thing-2"
      class="cdr-toggle-button__item"
      role="radio"
      tabindex="-1"
      value="thing-2"
    >
      
      thing-2
      
    </button>
  </li>
  <li
    class="cdr-toggle-button__container"
    role="presentation"
  >
    <button
      aria-checked="false"
      aria-label="thing-3"
      class="cdr-toggle-button__item"
      role="radio"
      tabindex="-1"
      value="thing-3"
    >
      
      thing-3
      
    </button>
  </li>
  
</ul>
`;

exports[`CdrToggleButton > component snapshot > matches the snapshot 1`] = `
<ul
  class="cdr-toggle-group cdr-toggle-group--medium"
  role="radiogroup"
>
  
  <li
    class="cdr-toggle-button__container"
    role="presentation"
  >
    <button
      aria-checked="true"
      aria-label="thing-1"
      class="cdr-toggle-button__item"
      role="radio"
      tabindex="0"
      value="thing-1"
    >
      
      thing-1
      
    </button>
  </li>
  <li
    class="cdr-toggle-button__container"
    role="presentation"
  >
    <button
      aria-checked="false"
      aria-label="thing-2"
      class="cdr-toggle-button__item"
      role="radio"
      tabindex="-1"
      value="thing-2"
    >
      
      thing-2
      
    </button>
  </li>
  <li
    class="cdr-toggle-button__container"
    role="presentation"
  >
    <button
      aria-checked="false"
      aria-label="thing-3"
      class="cdr-toggle-button__item"
      role="radio"
      tabindex="-1"
      value="thing-3"
    >
      
      thing-3
      
    </button>
  </li>
  
</ul>
`;

exports[`CdrToggleButton > component snapshot > when clicking on an unchecked button > matches snapshot 1`] = `
<ul
  class="cdr-toggle-group cdr-toggle-group--medium"
  role="radiogroup"
>
  
  <li
    class="cdr-toggle-button__container"
    role="presentation"
  >
    <button
      aria-checked="false"
      aria-label="thing-1"
      class="cdr-toggle-button__item"
      role="radio"
      tabindex="-1"
      value="thing-1"
    >
      
      thing-1
      
    </button>
  </li>
  <li
    class="cdr-toggle-button__container"
    role="presentation"
  >
    <button
      aria-checked="true"
      aria-label="thing-2"
      class="cdr-toggle-button__item"
      role="radio"
      tabindex="0"
      value="thing-2"
    >
      
      thing-2
      
    </button>
  </li>
  <li
    class="cdr-toggle-button__container"
    role="presentation"
  >
    <button
      aria-checked="false"
      aria-label="thing-3"
      class="cdr-toggle-button__item"
      role="radio"
      tabindex="-1"
      value="thing-3"
    >
      
      thing-3
      
    </button>
  </li>
  
</ul>
`;
